### PR TITLE
SDK-2466: Enhance upgrader logic with ability to use security rel…

### PR DESCRIPTION
**Pull request by VladislavStrelchenko**
Ticket: https://spryker.atlassian.net/browse/SDK-2928

### CODE REVIEW

File: src/ReleaseApp/Domain/Entities/UpgradeInstructionsReleaseGroup.php [hunk 1]
```diff
@@ -180,4 +185,12 @@ public function getJiraIssueLink(): ?string
     {
         return isset($this->body[static::JIRA_KEY]) ? $this->body[static::JIRA_KEY][static::ISSUE_LINK_KEY] : null;
     }
+
+    /**
+     * @return bool
+     */
+    public function isSecurity(): bool
+    {
+        return isset($this->body[static::SECURITY_KEY]) ? (bool)$this->body[static::SECURITY_KEY] : false;
```
Comment (sergeyspryker) [inline src/ReleaseApp/Domain/Entities/UpgradeInstructionsReleaseGroup.php:194]: Minor: (bool) ($this->body[static::SECURITY_KEY] ?? false)

File: src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php [hunk 2]
```diff
@@ -88,6 +88,30 @@ public function sortByReleasedDate(): self
         return new self(array_values($sortData));
     }
 
+    /**
+     * @return self
+     */
+    public function getOnlySecurityFixes(): self
```
Comment (sergeyspryker) [inline src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php:94]: getSecurityFixes

File: src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php [hunk 3]
```diff
@@ -88,6 +88,30 @@ public function sortByReleasedDate(): self
...
+            array_filter($this->elements, function ($releaseGroup) {
+                return $releaseGroup->isSecurity();
+            }),
+        );
+    }
+
+    /**
+     * @return self
+     */
+    public function filterSecurityFixes(): self
```
Comment (sergeyspryker) [inline src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php:106]: getNonSecurityFixes

File: src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php [hunk 4]
```diff
@@ -88,6 +88,30 @@ public function sortByReleasedDate(): self
         return new self(array_values($sortData));
     }
 
+    /**
+     * @return self
+     */
+    public function getOnlySecurityFixes(): self
+    {
+        return new self(
+            array_filter($this->elements, function ($releaseGroup) {
```
Comment (sergeyspryker) [inline src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php:97]: static fn(...): bool => $releaseGroup->isSecurity();

Add type-hints to the function

File: src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php [hunk 5]
```diff
@@ -88,6 +88,30 @@ public function sortByReleasedDate(): self
...
+        );
+    }
+
+    /**
+     * @return self
+     */
+    public function filterSecurityFixes(): self
+    {
+        return new self(
+            array_filter($this->elements, function ($releaseGroup) {
```
Comment (sergeyspryker) [inline src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php:109]: Same as above

File: src/ReleaseApp/Infrastructure/Shared/Dto/Collection/ReleaseGroupDtoCollection.php [hunk 6]
```diff
@@ -82,4 +82,16 @@ public function getCommonModuleCollection(): ModuleDtoCollection
 
         return $resultCollection;
     }
+
+    /**
+     * @return self
+     */
+    public function getOnlySecurityFixes(): self
```
Comment (sergeyspryker) [inline src/ReleaseApp/Infrastructure/Shared/Dto/Collection/ReleaseGroupDtoCollection.php:89]: Same as above

File: src/Upgrade/Application/Strategy/ReleaseApp/ReleaseGroupFilter/SecurityMajorFilterItem.php [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    /**
+     * @param string $semanticVersion
+     *
+     * @throws \RuntimeException
+     *
+     * @return int
+     */
+    protected function getMajorVersion(string $semanticVersion): int
```
Comment (sergeyspryker) [inline src/Upgrade/Application/Strategy/ReleaseApp/ReleaseGroupFilter/SecurityMajorFilterItem.php:69]: would be fine to move this method into the separate VersionHelper

File: src/Upgrade/Application/Strategy/ReleaseApp/ReleaseGroupFilter/SecurityMajorFilterItem.php [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        $versionParts = explode('.', $semanticVersion);
+        if (!$versionParts || !count($versionParts)) {
```
Comment (sergeyspryker) [inline src/Upgrade/Application/Strategy/ReleaseApp/ReleaseGroupFilter/SecurityMajorFilterItem.php:72]: I guess it always be false

File: src/Upgrade/Application/Strategy/ReleaseApp/ReleaseGroupFilter/SecurityMajorFilterItem.php [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            throw new RuntimeException(
+                sprintf('Unknown semantic version %s;', $semanticVersion),
+            );
+        }
+
+        $major = array_shift($versionParts);
+
+        if (!is_int($major) && !is_string($major)) {
```
Comment (sergeyspryker) [inline src/Upgrade/Application/Strategy/ReleaseApp/ReleaseGroupFilter/SecurityMajorFilterItem.php:80]: i guess major always be string

File: src/Upgrade/Application/Strategy/ReleaseApp/ReleaseGroupFilter/SecurityMajorFilterItem.php [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * @param string $semanticVersion
+     *
+     * @throws \RuntimeException
+     *
+     * @return int
+     */
+    protected function getMajorVersion(string $semanticVersion): int
+    {
```
Comment (sergeyspryker) [inline src/Upgrade/Application/Strategy/ReleaseApp/ReleaseGroupFilter/SecurityMajorFilterItem.php:70]: need to add regexp validation i hope

**Review by sergeyspryker — CHANGES_REQUESTED** [review]

File: src/ReleaseApp/Domain/Entities/UpgradeInstructionsReleaseGroup.php (hunk 1 above)
Comment (VladislavStrelchenko) [reply src/ReleaseApp/Domain/Entities/UpgradeInstructionsReleaseGroup.php:194]: ++

File: src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php (hunk 3 above)
Comment (VladislavStrelchenko) [reply src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php:106]: ++

File: src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php (hunk 2 above)
Comment (VladislavStrelchenko) [reply src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php:94]: ++

File: src/ReleaseApp/Domain/Entities/UpgradeInstructionsReleaseGroup.php (hunk 1 above)
Comment (VladislavStrelchenko) [reply src/ReleaseApp/Domain/Entities/UpgradeInstructionsReleaseGroup.php:194]: ++

File: src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php (hunk 4 above)
Comment (VladislavStrelchenko) [reply src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php:97]: ++

File: src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php (hunk 5 above)
Comment (VladislavStrelchenko) [reply src/ReleaseApp/Domain/Entities/Collection/UpgradeInstructionsReleaseGroupCollection.php:109]: ++

File: src/ReleaseApp/Infrastructure/Shared/Dto/Collection/ReleaseGroupDtoCollection.php (hunk 6 above)
Comment (VladislavStrelchenko) [reply src/ReleaseApp/Infrastructure/Shared/Dto/Collection/ReleaseGroupDtoCollection.php:89]: ++

File: src/Upgrade/Application/Strategy/ReleaseApp/ReleaseGroupFilter/SecurityMajorFilterItem.php (hunk 8 above)
Comment (VladislavStrelchenko) [reply src/Upgrade/Application/Strategy/ReleaseApp/ReleaseGroupFilter/SecurityMajorFilterItem.php:72]: Thanks updated

File: src/Upgrade/Application/Strategy/ReleaseApp/ReleaseGroupFilter/SecurityMajorFilterItem.php (hunk 10 above)
Comment (VladislavStrelchenko) [reply src/Upgrade/Application/Strategy/ReleaseApp/ReleaseGroupFilter/SecurityMajorFilterItem.php:70]: ++

**Review by sergeyspryker — APPROVED** [review]

File: src/Core/Infrastructure/SemanticVersionHelper.php [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    /**
+     * @param string $version
+     *
+     * @return bool
+     */
+    protected static function isSemanticVersion(string $version): bool
+    {
+        return preg_match('/^\d+\.\d+\.\d+$/', $version) === 1;
```
Comment (DmytroKlymanSpryker) [inline src/Core/Infrastructure/SemanticVersionHelper.php:40]: Semantic version can contain also the stability suffix - stable|RC|beta|alpha|dev . And it is used in Spryker, eg.g. propel package.

In this case regex won't work. I think it's better to use `composer/semver` if possible.

File: tests/Core/Infrastructure/SemanticVersionHelperTest.php [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+use Core\Infrastructure\SemanticVersionHelper;
+use PHPUnit\Framework\TestCase;
+
+class SemanticVersionHelperTest extends TestCase
+{
+    /**
+     * @dataProvider getMajorVersionProvider
+     *
+     * @param mixed $value
+     * @param ?int|int $expectedResult
```
Comment (DmytroKlymanSpryker) [inline tests/Core/Infrastructure/SemanticVersionHelperTest.php:21]: int|null

File: tests/Core/Infrastructure/SemanticVersionHelperTest.php [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+            ['foo-bar-baz', null],
+            [555, null],
+            [-555, null],
+            ['1.0.0', 1],
+            ['1.0.2', 1],
+            ['1.2.2', 1],
+            ['5.2.222', 5],
+            ['55.28888.222', 55],
+            ['-55.28888.222', null],
+        ];
```
Comment (DmytroKlymanSpryker) [inline tests/Core/Infrastructure/SemanticVersionHelperTest.php:46]: Here is few more examples of the semantic versions: 

Valid Semantic Versions:
1.0.0
0.1.0
0.0.1
10.20.30
2.3.4-beta
5.6.7-alpha.123
1.2.3-rc.456+build789
3.2.1+build123
4.5.6-beta.789+build999

Invalid Semantic Versions:
1.2
1.2.a
1.2.3.4
1.2.3-rc!
1.2.3-beta@456
1.2.3+build_789
1.2.3-rc.456+build_789
1.2.3-beta+build!999
1.2.3-beta+build.999-456

File: src/Core/Infrastructure/SemanticVersionHelper.php (hunk 11 above)
Comment (VladislavStrelchenko) [reply src/Core/Infrastructure/SemanticVersionHelper.php:40]: thanks replaced with `return Comparator::greaterThanOrEqualTo($version, '0.0.1') === true;`

File: tests/Core/Infrastructure/SemanticVersionHelperTest.php (hunk 13 above)
Comment (VladislavStrelchenko) [reply tests/Core/Infrastructure/SemanticVersionHelperTest.php:46]: Thanks, I have added the cases to test 
but most of Invalid versions `composer/semver` determinate as valid

**Review by DmytroKlymanSpryker — APPROVED** [review]

**Pull request merged**